### PR TITLE
[v6r15] Several fixes for storage plugins, especially gfal2

### PR DESCRIPTION
**Pull request by chaen**
This PR contains numerous fixes and improvements to the StorageElement plugins, in particular for gfal2. 
* Add a new way of doing pfnparse and pfnunparse using the standard python library. The two methods now contains a flag to know which method to use. By default, the old hand made one is used. The new one works perfectly for all standard protocols, except SRM. Of course, unit tests are there
* The mangling of the urls is now cleaner in the plugins, and is done using pfnparse/pfnunparse rather than appending/prepending strings to each others
* a New attribute is added to the storage plugins: DYNAMIC_OPTIONS. This allows to construct URLs with attributes going at the end of the URL, in the form ```?key1=value1&key2=value2``` This is useful for xroot and http.
* Improved system test for gfal2. 

This was tested with the help of @andresailer.
It is to be noted that there is a bug in castor preventing recursive directory creation for the moment with xroot, and I also spot a bug in the connection caching system of gfal2. We anyway will need to upgrade the library versions.

### CODE REVIEW

File: Core/Utilities/Pfn.py [hunk 1]
```diff
@@ -13,29 +13,44 @@
...
+  """
   Create PFN URI from pfnDict
 
-  :param dict pfnDict: 
+  :param dict pfnDict:
   """
-  ## make sure all keys are in
+  # # make sure all keys are in
   allDict = dict.fromkeys( [ "Protocol", "Host", "Port", "WSUrl", "Path", "FileName" ], "" )
   if type( allDict ) != type( pfnDict ):
```
Comment (fstagni) [inline Core/Utilities/Pfn.py]: use `isinstance`?

File: tests/Integration/Resources/Storage/Test_Resources_GFAL2StorageBase.py [hunk 2]
```diff
@@ -1,38 +1,63 @@
...
-gfal2-plugin-xrootd-0.4.0
-gfal2-python-1.8.3
-
-Storage element:
-Change self.storageName within the basicTest class to the name of the storageElement you wish to run this test against.
-This storage element needs to be supporting two access protocols, using the protocolName GFAL2_SRM2 and GFAL2_XROOT
-respectively. If you want to test only either SRM2 or XROOT remove the other testsuite in the unittest.
-
+This integration tests will perform basic operations on a storage element, depending on which protocols are available.
+It creates a local hierarchy, and then tries to upload, download, remove, get metadata etc
```
Comment (fstagni) [inline tests/Integration/Resources/Storage/Test_Resources_GFAL2StorageBase.py]: Can you provide an example on how to run this test?

Also, I guess it needs a "real" SE behind. So, a "real" certificate, correct?

Comment (chaen) [inline tests/Integration/Resources/Storage/Test_Resources_GFAL2StorageBase.py]: Yes, it is a system test, not an integration one

**Comment by atsareg** [comment]
Are you sure this is a patch level PR ?

**Comment by chaen** [comment]
I must confess I hesitated a bit, but since most of the behavior is unchanged due to the default flag, and that André and I tested quite extensively, I thought it could go. Moreover, this also contains fixes for XROOT plugin used by ILC.

File: Resources/Storage/GFAL2_StorageBase.py [hunk 3]
```diff
@@ -319,31 +334,31 @@ def __putSingleFile( self, src_file, dest_url, sourceSize ):
       # no checksum check, compare file sizes for verfication
       else:
         res = self.__getSingleFileSize( dest_url )
-        if res['OK']:
-          destSize = res['Value']
-        else:
-          destSize = None
-        self.log.debug( 'GFAL2_StorageBase.__putSingleFile: destSize: %s, sourceSize: %s' % ( destSize, sourceSize ) )
+        if not res['OK']:
+          return res
```
Comment (andresailer) [inline Resources/Storage/GFAL2_StorageBase.py]: If there is a return here the cleanup of the failed transfer will not happen, or will this always happen when an error is returned from this function?

File: Resources/Storage/GFAL2_StorageBase.py [hunk 4]
```diff
@@ -663,28 +681,35 @@ def __getSingleMetadata( self, path ):
...
     metadataDict = self.__parseStatInfoFromApiOutput( statInfo )
 
-    if metadataDict['File']:
-      if self.checksumType:
-        res = self.__getChecksum( path, self.checksumType )
-        metadataDict['Checksum'] = res.get( 'Value', '' )
+    if metadataDict['File'] and self.checksumType:
+      res = self.__getChecksum( path, self.checksumType )
+      if not res['OK']:
+        log.warning( "Could not get checksum:%s" % res['Message'] )
```
Comment (andresailer) [inline Resources/Storage/GFAL2_StorageBase.py]: `log.warn`

File: Resources/Storage/GFAL2_StorageBase.py [hunk 5]
```diff
@@ -663,28 +681,35 @@ def __getSingleMetadata( self, path ):
...
 
     metadataDict = self._addCommonMetadata( metadataDict )
 
-    attributeDict = self._getExtendedAttributes( path ).get( 'Value' , {} )
-    # add extended attributes to the dict if available
-    self._updateMetadataDict( metadataDict, attributeDict )
+    if self._defaultExtendedAttributes is not None:
+      res = self._getExtendedAttributes( path, attributes = self._defaultExtendedAttributes )
+      if not res['OK']:
+        log.warning( "Could not get extended attributes: %s" % res['Message'] )
```
Comment (andresailer) [inline Resources/Storage/GFAL2_StorageBase.py]: `log.warn`

File: Resources/Storage/GFAL2_StorageBase.py [hunk 6]
```diff
@@ -1297,29 +1365,30 @@ def __getSingleDirectory( self, src_dir, dest_dir ):
...
 
     filesReceived = 0
     sizeReceived = 0
 
     res = self.__isSingleDirectory( src_dir )
     if not res['OK']:
-      errStr = 'GFAL2_StorageBase.__getSingleDirectory: Failed to find the source directory'
-      self.log.debug( res['Message'], src_dir )
+      errStr = 'Failed to find the source directory'
+      log.debug( res['Message'], src_dir )
```
Comment (andresailer) [inline Resources/Storage/GFAL2_StorageBase.py]: `errStr` is not used here

File: Resources/Storage/GFAL2_XROOTStorage.py [hunk 7]
```diff
@@ -47,47 +49,35 @@ def __init__( self, storageName, parameters ):
...
-
-    :param self: self reference
-    :param str src_url: path of the source file
-    :param str dest_file: path of destination
-    :returns: S_ERROR( errStr ) in case of an error
-              S_OK( size of file ) if copying is successful
 
+  def __addDoubleSlash( self, res ):
+    """ Utilities to add the double slash between the host(:port) and the path
+        :param res: DIRAC return structure which contains an URL if S_OK
```
Comment (andresailer) [inline Resources/Storage/GFAL2_XROOTStorage.py]: newline before ":param" block

File: Resources/Storage/XROOTStorage.py [hunk 8]
```diff
@@ -1019,8 +1027,14 @@ def __getSingleDirectory( self, src_dir, dest_dir ):
     receivedAllFiles = True
     self.log.debug( "XROOTStorage.__getSingleDirectory: Trying to first download the %s files." % len( sFilesDict ) )
     for sFile in sFilesDict:
+      # Getting the last filename
+      res = pfnparse( sFile, srmSpecific = self.srmSpecificParse )
+      if not res['OK']:
+        self.log.error( "XROOTStorage.__getSingleDirectory: :Cannot unparse target file", res )
```
Comment (andresailer) [inline Resources/Storage/XROOTStorage.py]: Colon too much, before ":Cannot"

File: Resources/Storage/StorageBase.py [hunk 9]
```diff
@@ -317,17 +324,20 @@ def constructURLFromLFN( self, lfn, withWSUrl = False ):
...
-    result = self.getURLBase( withWSUrl = withWSUrl )
-    if not result['OK']:
-      return result
-    urlBase = result['Value']
-    url = os.path.join( urlBase, lfn.lstrip( '/' ) )
-    return S_OK( url )
+    urlDict = dict( self.protocolParameters )
+    urlDict['Options'] = '&'.join( "%s=%s" % ( optionName, urlDict[paramName] )
+                                   for paramName, optionName in self.DYNAMIC_OPTIONS.iteritems()
+                                    if urlDict.get( paramName ) )
```
Comment (andresailer) [inline Resources/Storage/StorageBase.py]: wrong indentation ?

File: Resources/Storage/GFAL2_StorageBase.py (hunk 3 above)
Comment (chaen) [inline Resources/Storage/GFAL2_StorageBase.py]: ahhh right, that was the reason for setting destSize to None ! I'll revert this change. Thanks !

File: Resources/Storage/StorageBase.py (hunk 9 above)
Comment (chaen) [inline Resources/Storage/StorageBase.py]: Really ? How would you want it indented ? 
The for loop is inside the join, so I thought it makes sense to allign it this way. How should it be ?

Comment (andresailer) [inline Resources/Storage/StorageBase.py]: the "if" has one to many spaces

Comment (chaen) [inline Resources/Storage/StorageBase.py]: Ah right I had missed that one ! Thanks

File: Resources/Storage/GFAL2_StorageBase.py [hunk 10]
```diff
@@ -396,31 +410,35 @@ def _getSingleFile( self, src_url, dest_file, disableChecksum = False ):
...
+      log.error( "Error while determining file size", res['Message'] )
       return res
 
     remoteSize = res['Value']
 
     # Set gfal2 copy parameters
     # folder is created and file exists, setting known copy parameters
     params = self.gfal2.transfer_parameters()
-    params.timeout = int( remoteSize / self.MIN_BANDWIDTH + 300 )
+    params.timeout = 10 + int( remoteSize / self.MIN_BANDWIDTH * 4 + 300 )
```
Comment (andresailer) [inline Resources/Storage/GFAL2_StorageBase.py]: Why this change? Why +10 and not + 310 instead? And the factor 4 is just to have longer timeout? This probably warrants a comment.

Comment (chaen) [inline Resources/Storage/GFAL2_StorageBase.py]: Frankly this is dark magic that was determined I believe empiricaly. I just made it consistent with what is in the original plugin, but in fact is is probably the opportunity to review it... maybe @phicharp knows better

Comment (chaen) [inline Resources/Storage/GFAL2_StorageBase.py]: So I just had a discussion with Philippe. He confirms that these are empirical values, and that probably the 10 and 300 corresponds to different overheads. However, I realized that these come from SRM plugin, but will now be shared by the different protocols. So I think I will just make a method to estimate the timeout, with this as default value (with a comment ;-) ) , so we can overwrite it for a given protocol if needed. Fine with you ?

Comment (andresailer) [inline Resources/Storage/GFAL2_StorageBase.py]: Yeah, fine by me. Just wondering about three (dark) magic numbers :)

Comment (chaen) [inline Resources/Storage/GFAL2_StorageBase.py]: In practice, we see that when there is a timeout, it really means sluggish link, so it should not be too bad :-) But I just sent a mail to Alejandro to see if these estimations are still good or need to be reviewed with gfal2

**Comment by fstagni** [comment]
IMHO Review OK

**Comment by atsareg** [comment]
Review OK
